macros/replacer_test.go

```diff
@@ -0,0 +1,115 @@
+	"http://pubmatic.com?macro1=##PBS-BIDID##&macro2=##PBS-APPBUNDLE##&macro3=##PBS-APPBUNDLE##&macro4=##PBS-PUBDOMAIN##&macro5=##PBS-PAGEURL##&macro6=##PBS-ACCOUNTID##&macro6=##PBS-LIMITADTRACKING##&macro7=##PBS-GDPRCONSENT##&macro8=##PBS-GDPRCONSENT##&macro9=##PBS-MACRO-CUSTOMMACR1CUST1##&macro10=##PBS-MACRO-CUSTOMMACR1CUST2##",
+	"http://testbidder.com?macro1=##PBS-BIDID##&macro2=##PBS-APPBUNDLE##&macro3=##PBS-APPBUNDLE##&macro4=##PBS-PUBDOMAIN##&macro5=##PBS-PAGEURL##&macro6=##PBS-ACCOUNTID##&macro6=##PBS-LIMITADTRACKING##&macro7=##PBS-GDPRCONSENT##&macro8=##PBS-GDPRCONSENT##&macro9=##PBS-MACRO-CUSTOMMACR1CUST1##&macro10=##PBS-MACRO-CUSTOMMACR1CUST2##",
+	"http://dummybidder.com?macro1=##PBS-BIDID##&macro2=##PBS-APPBUNDLE##&macro3=##PBS-APPBUNDLE##&macro4=##PBS-PUBDOMAIN##&macro5=##PBS-PAGEURL##&macro6=##PBS-ACCOUNTID##&macro6=##PBS-LIMITADTRACKING##&macro7=##PBS-GDPRCONSENT##&macro8=##PBS-GDPRCONSENT##&macro9=##PBS-MACRO-CUSTOMMACR1CUST1##&macro10=##PBS-MACRO-CUSTOMMACR1CUST2##",
+}
+
+var req *openrtb_ext.RequestWrapper = &openrtb_ext.RequestWrapper{
+	BidRequest: &openrtb2.BidRequest{
+		ID: "123",
+		Site: &openrtb2.Site{
+			Domain: "testdomain",
+			Publisher: &openrtb2.Publisher{
+				Domain: "publishertestdomain",
+				ID:     "testpublisherID",
+			},
+			Page: "pageurltest",
+		},
+		App: &openrtb2.App{
+			Domain: "testdomain",
+			Bundle: "testbundle",
+			Publisher: &openrtb2.Publisher{
+				Domain: "publishertestdomain",
+				ID:     "testpublisherID",
+			},
+		},
+		Device: &openrtb2.Device{
+			Lmt: &lmt,
+		},
+		User: &openrtb2.User{Ext: []byte(`{"consent":"yes" }`)},
+		Ext:  []byte(`{"prebid":{"channel": {"name":"test1"},"macros":{"CUSTOMMACR1":"value1","CUSTOMMACR2":"value2","CUSTOMMACR3":"value3"}}}`),
+	},
+}
+
+var bid *openrtb2.Bid = &openrtb2.Bid{ID: "bidId123", CID: "campaign_1", CrID: "creative_1"}
+
+func BenchmarkStringIndexCachedBasedReplacer(b *testing.B) {
+
+	processor := NewReplacer()
+	for n := 0; n < b.N; n++ {
+		for _, url := range benchmarkURL {
+			macroProvider := NewProvider(req)
+			macroProvider.PopulateBidMacros(&entities.PbsOrtbBid{Bid: bid}, "test")
+			macroProvider.PopulateEventMacros("123", "tracking", "firstQuartile")
+			_, err := processor.Replace(url, macroProvider)
+			if err != nil {
+				b.Errorf("Fail to replace macro in tracker")
+			}
+		}
+	}
+}
+
+// func BenchmarkGolangReplacer(b *testing.B) {
+
+// 	for n := 0; n < b.N; n++ {
+// 		for _, url := range benchmarkURL {
+// 			macroProvider := NewProvider(req)
+
+// 			macroProvider.SetContext(MacroContext{
+// 				Bid:            &entities.PbsOrtbBid{Bid: bid},
+// 				Imp:            nil,
+// 				Seat:           "test",
+// 				VastCreativeID: "123",
+// 				VastEventType:  "firstQuartile",
+// 				EventElement:   "tracking",
+// 			})
+// 			StringReplacer(url, macroProvider)
+// 		}
+// 	}
+// }
```

Comment:
Thanks so much for benchmarking the different approaches and sharing the results. We really appreciate the data driven decision :)

However, can we please move this info in the comments section of the PR and remove it from this file?

Reply:
Done